### PR TITLE
Add route tests and reorganize routes

### PR DESCRIPTION
**Pull request by zaucker**
Parameters might not all be correct yet. Will be fixed in next PRs together with functional tests for Agrammon::Web::Service.

### CODE REVIEW

File: dev/META6.json [hunk 1]
```diff
@@ -11,12 +11,15 @@
...
     "Cro::HTTP",
     "Cro::HTTP::Session::Pg",
+    "Cro::HTTP::Test",
     "DB::Pg",
     "Digest::SHA1::Native",
     "OO::Monitors",
     "Shell::Command",
     "Spreadsheet::XLSX",
     "Text::CSV",
+    "Test",
```
Comment (jnthn) [inline dev/META6.json:21]: Hm, `Test` is provided in core, so it isn't usually mentioned in the `depends`.

File: dev/META6.json [hunk 2]
```diff
@@ -11,12 +11,15 @@
   "depends": [
     "Cro::HTTP",
     "Cro::HTTP::Session::Pg",
+    "Cro::HTTP::Test",
```
Comment (jnthn) [inline dev/META6.json:14]: Not a big issue, but strictly belongs in `test-depends`.

File: lib/Agrammon/Web/Routes.pm6 [hunk 3]
```diff
@@ -5,100 +5,67 @@ use Cro::HTTP::Router;
 use Agrammon::Web::Service;
 use Agrammon::Web::SessionUser;
 
-
 sub routes(Agrammon::Web::Service $ws) is export {
+    subset LoggedIn of Agrammon::Web::SessionUser where .logged-in;
```
Comment (jnthn) [inline lib/Agrammon/Web/Routes.pm6:9]: It's clearer to put this entirely outside of `routes`, at the top level, given it's package-scoped.

File: lib/Agrammon/Web/Routes.pm6 [hunk 4]
```diff
@@ -5,100 +5,67 @@ use Cro::HTTP::Router;
...
+
+    # Fix for Agrammon
+    # my $root = $app ~~ Easii::Application::Development ?? 'frontend/source-output' !! 'qx-build';
```
Comment (jnthn) [inline lib/Agrammon/Web/Routes.pm6:12]: Maybe don't mention Easii? :-) Also, I'm curious if `$root` is over-used in the static content? It seems primarily relevant for the scripts, but maybe not the HTML?

File: lib/Agrammon/Web/Routes.pm6 [hunk 5]
```diff
@@ -213,24 +172,82 @@ sub routes(Agrammon::Web::Service $ws) is export {
...
-            request-body -> (:$dataset!) {
-                my $data = $ws.get-input($user, $dataset);
+sub user-routes(Agrammon::Web::Service $ws) {
+    route {
+
+        # implement/test
+        post -> LoggedIn $user, 'change_password' {
+            request-body -> (:$old-password!, :$new-password!) {
+                my $data = $ws.change-password($old-password, $new-password);
+                content 'application/json', $data;
```
Comment (jnthn) [inline lib/Agrammon/Web/Routes.pm6:190]: Looking at this made me a bit curious: how does `$ws` know which user it is to change the password for? Should we be passing something to it? Same for various other operations.

File: lib/Agrammon/Web/Service.pm6 [hunk 6]
```diff
@@ -84,8 +119,15 @@ class Agrammon::Web::Service {
...
+        ...
+        $user.change-password($oldPassword, $newPassword);
+        return $user;
+    }
+
+    method reset-password(Agrammon::Web::SessionUser $user, Str $email, Str $password, Str $key) {
+        ...
+        $user.reset-password($email, $password);
         return $user;
     }
```
Comment (jnthn) [inline lib/Agrammon/Web/Service.pm6:132]: Here I see it *does* expect to be passed the user. So I guess the route code needs updating.

File: t/routes-application.t [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    }
+}
+
+done-testing;
+
+=begin pod
+
+=COPYRIGHT Copyright (c) 2020 by OETIKER+PARTNER AG. All rights reserved.
+
+=AUTHOR S<Fritz WorthZauckerington E<lt>fritz.zaucker@oetiker.chE<gt>>
```
Comment (jnthn) [inline t/routes-application.t:191]: Who? :rofl:

File: t/routes-application.t [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+# my $fake-user = %( :id(1), :username('foor@bar.ch'), );
+
+# my $fake-tag-a = %( :id(42), :name('TagA'), :user($fake-user));
+# my $fake-tag-b = %( :id(43), :name('TagB'), :user($fake-user));
+
+# my $fake-dataset-a = %( :id(42), :name('DatasetA'), :user($fake-user));
+# my $fake-dataset-b = %( :id(43), :name('DatasetB'), :user($fake-user));
+
```
Comment (jnthn) [inline t/routes-application.t:25]: Maybe not applicable in this test?

File: lib/Agrammon/Web/Service.pm6 (hunk 6 above)
Comment (jnthn) [reply lib/Agrammon/Web/Service.pm6:132]: Ah, now I see the "Parameters might not all be correct yet" mention at the top of the PR, maybe it's about this.

File: dev/META6.json (hunk 1 above)
Comment (zaucker) [reply dev/META6.json:21]: Removed

File: dev/META6.json (hunk 2 above)
Comment (zaucker) [reply dev/META6.json:14]: Moved together with the other test related modules.

File: lib/Agrammon/Web/Routes.pm6 (hunk 3 above)
Comment (zaucker) [reply lib/Agrammon/Web/Routes.pm6:9]: Moved

File: lib/Agrammon/Web/Routes.pm6 (hunk 4 above)
Comment (zaucker) [reply lib/Agrammon/Web/Routes.pm6:12]: Removed Easii ... would like to keep `$root` as is until I have the frontend build (for both local and deployed) figured out.

File: t/routes-application.t (hunk 7 above)
Comment (zaucker) [reply t/routes-application.t:191]: OMG ... fixed.

File: t/routes-application.t (hunk 8 above)
Comment (zaucker) [reply t/routes-application.t:25]: Removed

File: lib/Agrammon/Web/Routes.pm6 (hunk 5 above)
Comment (zaucker) [reply lib/Agrammon/Web/Routes.pm6:190]: In `Agrammon::Web::SessionUser` I do set the `username` attribute to the login user.
What if I also set the `id` attribute as well? Then I can use the SessionUser everywhere I need the logged-in user.

Or is this a bad design?

File: lib/Agrammon/Web/Service.pm6 (hunk 6 above)
Comment (zaucker) [reply lib/Agrammon/Web/Service.pm6:132]: See above.

File: lib/Agrammon/Web/Routes.pm6 [hunk 9]
```diff
@@ -185,8 +185,8 @@ sub user-routes(Agrammon::Web::Service $ws) {
 
         # implement/test
         post -> LoggedIn $user, 'change_password' {
-            request-body -> (:$oldPassword!, :$newPassword!) {
-                my $data = $ws.change-password($oldPassword, $newPassword);
+            request-body -> (:$old-password!, :$new-password!) {
```
Comment (jnthn) [inline lib/Agrammon/Web/Routes.pm6:188]: You're fine with the JSON object we get having to have the hyphens in them? (If not, you can do `:oldPassword($old-password)` to use a Raku-style variable name but extract it from another key in the JSON).

File: lib/Agrammon/Web/Routes.pm6 (hunk 5 above)
Comment (jnthn) [reply lib/Agrammon/Web/Routes.pm6]: No, that seems sensible enough to me. I mean, carrying around huge amounts of user data in there can make the session expensive to save/load (though we can also lazily populate it), but I don't think one extra ID attribute is a problem at all.

File: lib/Agrammon/Web/Routes.pm6 (hunk 9 above)
Comment (zaucker) [reply lib/Agrammon/Web/Routes.pm6:189]: I guess I overdid the Rakuization :-)

File: lib/Agrammon/Web/Routes.pm6 (hunk 5 above)
Comment (zaucker) [reply lib/Agrammon/Web/Routes.pm6]: As it is currently implemented, upon authentication **all** user info is loaded into the SessionUser object.
As it is not all that much info and we don't have large numbers of users, I think we can leave it like that for the moment.

File: lib/Agrammon/Web/Service.pm6 (hunk 6 above)
Comment (zaucker) [reply lib/Agrammon/Web/Service.pm6:132]: > Ah, now I see the "Parameters might not all be correct yet" mention at the top of the PR, maybe it's about this.

Actually I meant that I have to bring the parameters the frontend sends in sync with what the backend expects. This is probably correct for most routes, but I will have to verify that and in some cases the parameters might be a bit strangely named.

However, I found it easier to make the backend somewhat consistent and add route tests for it first, then in the next PR add the missing webservice calls (with tests) and then in the third step bring the frontend in sync.

File: lib/Agrammon/Web/Routes.pm6 (hunk 5 above)
Comment (zaucker) [reply lib/Agrammon/Web/Routes.pm6]: With that setup, method calls on the SessionUser object will automagically work on the right database entry.

**Review by jnthn — APPROVED** [review]

File: lib/Agrammon/Web/Routes.pm6 [hunk 10]
```diff
@@ -287,27 +302,24 @@ sub routes(Agrammon::Web::Service $ws) is export {
...
             }
         }
 
         # test/implement
-        post -> LoggedIn $user, 'order_instances', $name {
-            ...
-            request-body -> (:$name!) {
-                my $data = $ws.order-instances($user, $name);
+        post -> LoggedIn $user, 'order_instances' {
+            request-body -> (:@instances, :$dataset-name!) {
```
Comment (jnthn) [inline lib/Agrammon/Web/Routes.pm6:321]: Probably `:datasetName($dataset-name)`?

File: lib/Agrammon/Web/Routes.pm6 [hunk 11]
```diff
@@ -287,27 +302,24 @@ sub routes(Agrammon::Web::Service $ws) is export {
 
         # test/implement
         post -> LoggedIn $user, 'store_branch_data' {
-                ...
-            request-body -> (:$name!) {
-                my $data = $ws.store-branch-data($user, $name);
+            request-body -> (:%data!, :$dataset-name!) {
```
Comment (jnthn) [inline lib/Agrammon/Web/Routes.pm6:305]: `:datasetName($dataset-name)`

**Pull request merged**